File: cpbook/3_problem_solving_paradigms/11100_the_trip_2007.cpp
// UVa 11100 - The Trip, 2007
// https://onlinejudge.org/external/111/11100.pdf

#include <iostream>
#include <vector>
#include <map>
#include <algorithm>

using namespace std;

int calculate_repeated(const vector<int>& bags)
{
	int result = 0;
	map<int, int> repeated;

	for (auto b : bags)	{
		repeated[b]++;
	}
	for (auto it = repeated.begin(); it != repeated.end(); it++)	{
		result = max(result, it->second);
	}
	return result;
}

void calculate_min(int n, vector<int>& bags, int& k, vector<vector<int>>& pieces)
{
	k = calculate_repeated(bags);
	pieces.clear();
	for (int i = 0; i < k; i++)	{
		vector<int> p;
		pieces.push_back(p);
	}
	sort(bags.begin(), bags.end());
	for (int i = 0; i < n; i += k)	{
		int j = 0;
		while (j < k && i + j < n)	{
			pieces[j].push_back(bags[i + j]);
			j++;
		}
	}
}

int main()
{
	bool first_case = true;
	int n, b, k;
	vector<int> bags;
	vector<vector<int>> pieces;

	ios::sync_with_stdio(0);
	cin.tie(0);
	while (cin >> n)	{
		if (n == 0)	{
			break;
		}
		if (first_case)	{
			first_case = false;
		}	else	{
			cout << "\n";
		}
		bags.clear();
		for (int i = 0; i < n; i++)	{
			cin >> b;
			bags.push_back(b);
		}
		calculate_min(n, bags, k, pieces);
		cout << k << "\n";
		for (auto piece : pieces)	{
			for (int i = 0; i < piece.size() - 1; i++)	{
				cout << piece[i] << " ";
			}
			cout << piece.back() << "\n";
		}
	}
	return 0;
}

File: cpbook/3_problem_solving_paradigms/11100_the_trip_2007.py
# UVa 11100 - The Trip, 2007
# https://onlinejudge.org/external/111/11100.pdf


def calculate_repeated(bags):
	repeated = {}
	for b in bags:
		if b not in repeated:
			repeated[b] = 1
		else:
			repeated[b] += 1
	result = 0
	for k in repeated:
		result = max(result, repeated[k])
	return result

def calculate_minimum(n, bags):
	k = calculate_repeated(bags)
	pieces = []
	for _ in range(k):
		pieces.append([])
	bags.sort()
	# print(k, pieces, n, bags)
	for i in range(0, n, k):
		# print(f"i={i}")
		j = 0
		while j < k and i + j < n:
			# print(f"\tj={j}")
			pieces[j].append(bags[i + j])
			j += 1
	return k, pieces


if __name__ == "__main__":
	first_case = True
	while True:
		n = int(input())
		if n == 0:
			break
		if first_case:
			first_case = False
		else:
			print()		# empty line between cases
		bags = [int(b) for b in input().split()]
		k, pieces = calculate_minimum(n, bags)
		print(k)
		for piece in pieces:
			for i in range(len(piece) - 1):
				print(f"{piece[i]} ", end="")
			print(piece[-1])
